raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java

```diff
@@ -17,218 +17,67 @@
 
     /**
      * Returns the end offset and epoch for the snapshot.
      */
-    public OffsetAndEpoch snapshotId() {
-        return snapshot.snapshotId();
-    }
+    OffsetAndEpoch snapshotId();
 
     /**
      * Returns the last log offset which is represented in the snapshot.
      */
-    public long lastContainedLogOffset() {
-        return snapshot.snapshotId().offset - 1;
-    }
+    long lastContainedLogOffset();
 
     /**
      * Returns the epoch of the last log offset which is represented in the snapshot.
      */
-    public int lastContainedLogEpoch() {
-        return snapshot.snapshotId().epoch;
-    }
+    int lastContainedLogEpoch();
 
     /**
      * Returns true if the snapshot has been frozen, otherwise false is returned.
-     *
+     * <p>
      * Modification to the snapshot are not allowed once it is frozen.
      */
-    public boolean isFrozen() {
-        return snapshot.isFrozen();
-    }
+    boolean isFrozen();
 
     /**
      * Appends a list of values to the snapshot.
-     *
+     * <p>
      * The list of record passed are guaranteed to get written together.
      *
      * @param records the list of records to append to the snapshot
      * @throws IllegalStateException if append is called when isFrozen is true
      */
-    public void append(List<T> records) {
-        if (snapshot.isFrozen()) {
-            String message = String.format(
-                "Append not supported. Snapshot is already frozen: id = '%s'.",
-                snapshot.snapshotId()
-            );
-
-            throw new IllegalStateException(message);
-        }
-
-        accumulator.append(snapshot.snapshotId().epoch, records);
-
-        if (accumulator.needsDrain(time.milliseconds())) {
-            appendBatches(accumulator.drain());
-        }
-    }
+    void append(List<T> records);
 
     /**
      * Freezes the snapshot by flushing all pending writes and marking it as immutable.
-     *
+     * <p>
      * Also adds a {@link SnapshotFooterRecord} to the end of the snapshot
      */
-    public void freeze() {
-        finalizeSnapshotWithFooter();
-        appendBatches(accumulator.drain());
-        snapshot.freeze();
-        accumulator.close();
-    }
+    void freeze();
 
     /**
      * Closes the snapshot writer.
      *
      * If close is called without first calling freeze the snapshot is aborted.
      */
-    public void close() {
-        snapshot.close();
-        accumulator.close();
-    }
+    void close();
 
-    private void appendBatches(List<CompletedBatch<T>> batches) {
-        try {
-            for (CompletedBatch<T> batch : batches) {
-                snapshot.append(batch.data);
-            }
-        } finally {
-            batches.forEach(CompletedBatch::release);
-        }
-    }
+    /**
+     * Adds a {@link SnapshotHeaderRecord} to snapshot
+     *
+     * @throws IllegalStateException if the snapshot is not empty
+     */
+    void initializeSnapshotWithHeader();
```

Comment:
This method was originally `private` method. Why should we expose it as an interface method?